src/util-interval-tree.c

```diff
@@ -0,0 +1,421 @@
+        SCLogDebug("range [%u:%u] overlaps tree range [%u:%u]", port, port2, ptr->port, ptr->port2);
+        return true;
+    }
+
+#if 0
+    if ((port < ptr->port2) && (ptr->port < port2)) {
+        SCLogDebug("Found overlap with [%d, %d] vs [%u:%u]", ptr->port, ptr->port2, port, port2);
+        return true;
+    }
+#endif
+    SCLogDebug("No overlap found for [%d, %d] w [%d, %d]", port, port2, ptr->port, ptr->port2);
+    return false;
+}
+
+void PrintSigGroupHeadSigs(DetectEngineCtx *de_ctx, SigGroupHead *sgh);
+void PrintSigGroupHeadSigs(DetectEngineCtx *de_ctx, SigGroupHead *sgh)
+{
+    for (uint32_t sig = 0; sig < sgh->init->max_sig_id + 1; sig++) {
+        if (sgh->init->sig_array[sig / 8] & (1 << (sig % 8))) {
+            SCLogDebug("sig %u enabled", de_ctx->sig_array[sig]->id);
+        }
+    }
+}
+
+void printIT_SigGroupHeadSigs(DetectEngineCtx *de_ctx, SigGroupHead *sgh);
+void printIT_SigGroupHeadSigs(DetectEngineCtx *de_ctx, SigGroupHead *sgh)
+{
+    for (uint32_t sig = 0; sig < sgh->init->max_sig_id + 1; sig++) {
+        if (sgh->init->sig_array[sig / 8] & (1 << (sig % 8))) {
+            printf(" %u", de_ctx->sig_array[sig]->id);
+        }
+    }
+}
+
+int COUNT = 4;
+void printIT(DetectEngineCtx *de_ctx, SCIntervalNode *node, int space,
+        void (*Print)(DetectEngineCtx *, SigGroupHead *));
+void printIT(DetectEngineCtx *de_ctx, SCIntervalNode *node, int space,
+        void (*Print)(DetectEngineCtx *, SigGroupHead *))
+{
+    // Base case
+    if (node == NULL)
+        return;
+
+    // Increase distance between levels
+    space += COUNT;
+
+    // Process right child first
+    printIT(de_ctx, IRB_RIGHT(node, irb), space, Print);
+
+    // Print current node after space
+    // count
+    printf("\n\n");
+    for (int i = COUNT; i < space; i++)
+        printf(" ");
+    printf("[%d, %d]: %s; p = %d; max = %d, sids:", node->port, node->port2,
+            IRB_COLOR(node, irb) ? "RED" : "BLACK",
+            IRB_PARENT(node, irb) ? IRB_PARENT(node, irb)->port : 0, node->max);
+    Print(de_ctx, node->sh);
+    printf("\n");
+
+    // Process left child
+    printIT(de_ctx, IRB_LEFT(node, irb), space, Print);
+}
+
+/**
+ * \brief Function to find all the overlaps of given ports with the existing
+ *        port ranges in the interval tree
+ * \param de_ctx Detection Engine Context
+ * \param port Given low port
+ * \param port2 Given high port
+ * \param ptr Pointer to the root of the tree
+ * \param list A list of DetectPort objects to be filled
+ */
+static void FindOverlaps(DetectEngineCtx *de_ctx, uint16_t port, uint16_t port2,
+        SCIntervalNode *root, DetectPort **list)
+{
+    SCLogDebug("\nfind overlaps for [%u:%u]", port, port2);
+    SCIntervalNode *ptr = root;
+    DetectPort *new_port = NULL;
+    int cnt_overlaps = 0;
+#if 0
+    printIT(de_ctx, root, 1, printIT_SigGroupHeadSigs);
+#endif
+    SCIntervalNode *stack[100], *current = root;
```

Comment:
todo: dynamic size, may need heap alloc

Reply:
I am having trouble figuring out the cap on the memory required for the stack. What did you have in mind?
We'll never enter the worse case which is memory complexity of O(n) as we're using red black trees.

Reply:
does the algo have a way to determine the longest branch of the tree?

Reply:
I couldn't find or determine a way for interval trees in particular but a red black tree can at most have a height of `2lg(n + 1)`.. where `n` is the number of nodes.